compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/MemoizedInlineClassReplacements.kt

```diff
@@ -149,7 +150,7 @@ class MemoizedInlineClassReplacements {
     }
 
     private fun buildReplacement(function: IrFunction, body: IrFunctionImpl.() -> Unit) =
-        buildFun {
+        buildFunWithDescriptor(function.descriptor) {
```

Comment:
Could you clarify why descriptor is reuqired here?

Reply:
I changed the function to use a more descriptive name and added a comment to explain what the descriptor does.

Reply:
@sfs Thanks for renaming!

But it's still not clear for me why descriptor should be passed and its container source should be keept.
Could it be fixed purely with ir approach? 
Maybe same JvmInlineClassLowering should be execute  for external declarations (at least for their inline parts, as we do in InterfaceDelegationLowering.generateDelegationToDefaultImpl: default method stubs are also constructed for external declarations without actual lowering of external declaration)? cc @udalov

Reply:
The culprit here is `InlineCodegen`, which relies on descriptors to recognize external declarations (see [InlineCodegen.kt#L557](https://github.com/JetBrains/kotlin/blob/728459a5046759a91075f3e03dba21076133feb5/compiler/backend/src/org/jetbrains/kotlin/codegen/inline/InlineCodegen.kt#L557)).

Alternatively we could override this method in IrInlineCodegen, but this would end up copying some code right now.

The inline class lowering already executes (lazily) for external declarations.

Reply:
@sfs Actually I would prefer to avoid descriptor usage cause at some point we want to get rid  descriptors off. But this work seems would depend on https://upsource.jetbrains.com/kotlin/review/KOTLIN-CR-3069 cause otherwise it would be impossible to extract proper facade name

Reply:
And from another point this will allow us to support another cases of inlining from binaries (especially to support CompileKotlinAgainstInlineKotlinTestGenerated in Ir in future)

Reply:
I don't think the usage of descriptors here is such a problem - the changes will simply become obsolete once IrInlineCodegen no longer depends on descriptors.

In the meantime though, this PR allows us to start fixing tests that rely on inlined external declaration with inline class arguments (such as all coroutine tests using `Result<T>`) and to actually use inline classes and inline functions together with the IR backend.

Reply:
@sfs After short investigation I've realized that this work will require new type mapper so pr could be merged as is. And further changes in inliner I will made by myself